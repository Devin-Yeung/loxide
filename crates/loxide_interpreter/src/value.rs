#[derive(Debug, Clone, PartialEq)]
pub enum Value {
    Number(f64),
    String(String),
    Boolean(bool),
    Nil,
}

impl Value {
    pub fn is_number(&self) -> bool {
        matches!(self, Value::Number(_))
    }

    pub fn is_string(&self) -> bool {
        matches!(self, Value::String(_))
    }

    pub fn is_boolean(&self) -> bool {
        matches!(self, Value::Boolean(_))
    }

    pub fn is_nil(&self) -> bool {
        matches!(self, Value::Nil)
    }

    pub fn as_number(&self) -> Option<f64> {
        match self {
            Value::Number(v) => Some(*v),
            _ => None,
        }
    }

    pub fn as_boolean(&self) -> Option<bool> {
        match self {
            Value::Boolean(v) => Some(*v),
            _ => None,
        }
    }
}
