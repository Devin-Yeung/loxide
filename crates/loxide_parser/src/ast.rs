#[derive(Debug)]
pub struct Expr<'src> {
    pub kind: ExprKind<'src>,
}

#[derive(Debug)]
pub enum ExprKind<'src> {
    Literal(Literal<'src>),
    Unary(UnaryExpr<'src>),
    Binary(BinaryExpr<'src>),
    Grouped(Box<Expr<'src>>),
    Variable(Variable<'src>),
    Assign(AssignExpr<'src>),
}

#[derive(Debug)]
pub enum Literal<'src> {
    String(std::borrow::Cow<'src, str>),
    Number(f64),
    Boolean(bool),
    Nil,
}

#[derive(Debug)]
pub enum UnaryOperator {
    Minus,
    Bang,
}

#[derive(Debug)]
pub struct UnaryExpr<'src> {
    pub operator: UnaryOperator,
    pub expr: Box<Expr<'src>>,
}

#[derive(Debug)]
pub enum BinaryOperator {
    EqualEqual,
    BangEq,
    GreaterEqual,
    Greater,
    Less,
    LessEqual,
    Minus,
    Plus,
    Slash,
    Star,
}

#[derive(Debug)]
pub struct BinaryExpr<'src> {
    pub lhs: Box<Expr<'src>>,
    pub rhs: Box<Expr<'src>>,
    pub operator: BinaryOperator,
}

#[derive(Debug)]
pub struct GroupedExpr<'src> {
    pub expr: Box<Expr<'src>>,
}

#[derive(Debug)]
pub struct AssignExpr<'src> {
    pub name: Variable<'src>,
    pub value: Box<Expr<'src>>,
}

impl<'src> AssignExpr<'src> {
    pub fn new(name: &'src str, value: Expr<'src>) -> Self {
        AssignExpr {
            name: Variable { name },
            value: Box::new(value),
        }
    }
}

#[derive(Debug)]
pub struct Variable<'src> {
    pub name: &'src str,
}

#[derive(Debug)]
pub struct ConditionStmt<'src> {
    pub condition: Expr<'src>,
    pub then_branch: Box<Stmt<'src>>,
    pub else_branch: Option<Box<Stmt<'src>>>,
}

#[derive(Debug)]
pub struct WhileStmt<'src> {
    pub condition: Expr<'src>,
    pub body: Box<Stmt<'src>>,
}

#[derive(Debug)]
#[non_exhaustive]
pub enum Stmt<'src> {
    Expression(Expr<'src>),
    PrintStmt(Expr<'src>),
    VarDeclaration(Variable<'src>, Option<Expr<'src>>),
    Block(Vec<Stmt<'src>>),
    Condition(ConditionStmt<'src>),
    While(WhileStmt<'src>),
}
