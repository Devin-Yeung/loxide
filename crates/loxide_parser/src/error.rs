use thiserror::Error;

#[derive(Error, Debug, Eq, PartialEq, Copy, Clone)]
pub enum SyntaxError {
    #[error("Unexpected character at line {0}")]
    UnexpectedChar(usize),
    #[error("Unexpected EOF")]
    UnexpectedEOF,
    #[error("Expected")]
    Expect(&'static str),
    #[error("Invalid Conversion")]
    InvalidConversion,
    #[error("Invalid Number")]
    InvalidNumber,
    #[error("Invalid string literal")]
    InvalidStringLiteral,
    #[error("Invalid keyword")]
    InvalidKeywords,
}
