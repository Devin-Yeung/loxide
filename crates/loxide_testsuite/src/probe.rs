use once_cell::sync::Lazy;
use std::backtrace::Backtrace;
use std::collections::HashMap;
use std::fmt::{Debug, Formatter};
use std::sync::Mutex;

static PROBE: Lazy<Mutex<Probe>> = Lazy::new(|| Mutex::new(Probe::new()));
#[derive(Debug)]
pub struct Probe {
    mapping: HashMap<String, Vec<FootPrint>>,
}

#[derive(Clone, PartialEq)]
pub struct FootPrint(String);

impl PartialEq<FootPrint> for &str {
    fn eq(&self, other: &FootPrint) -> bool {
        self.eq(&other.0)
    }
}

impl PartialEq<&str> for FootPrint {
    fn eq(&self, other: &&str) -> bool {
        self.0.eq(other)
    }
}

impl Debug for FootPrint {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.write_str(&self.0)
    }
}

impl Probe {
    fn new() -> Self {
        Self {
            mapping: HashMap::new(),
        }
    }

    pub fn register(test_name: &str) {
        PROBE
            .lock()
            .unwrap()
            .mapping
            .insert(test_name.to_string(), Vec::new());
    }

    pub fn probe(content: String) {
        let raw = Backtrace::capture().to_string();
        let backtrace = Self::backtrace(&raw);
        let mut guard = PROBE.lock().unwrap();
        for (k, v) in guard.mapping.iter_mut() {
            if backtrace.iter().any(|func| func == k) {
                v.push(FootPrint(content));
                break;
            }
        }
    }

    pub fn footprints(test_name: &str) -> Vec<FootPrint> {
        let guard = PROBE.lock().unwrap();
        guard.mapping.get(test_name).unwrap_or(&Vec::new()).clone()
    }

    fn backtrace(backtrace: &str) -> Vec<&str> {
        let mut stacks = Vec::new();
        for line in backtrace.lines() {
            let trimmed = line.trim_start();
            if &trimmed[..2] != "at" {
                let func = trimmed[trimmed.find(':').unwrap() + 1..].trim();
                stacks.push(func);
            }
        }
        stacks
    }
}

#[cfg(test)]
mod tests {
    use crate::{footprints, probe, register};

    fn nested(num: i32) {
        outer(num);
    }

    fn outer(num: i32) {
        inner(num);
    }

    fn inner(num: i32) {
        probe!(num);
    }

    #[test]
    fn same_prefix() {
        fn same_prefix_foo() {
            register!();
            nested(42);
            assert_eq!("42", footprints!()[0]);
        }

        fn same_prefix_bar() {
            register!();
            nested(-42);
            assert_eq!(footprints!()[0], "-42");
        }
        same_prefix_foo();
        same_prefix_bar();
    }
}
