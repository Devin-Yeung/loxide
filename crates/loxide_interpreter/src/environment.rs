use crate::error::RuntimeError;
use crate::value::Value;
use std::collections::HashMap;

pub struct Environment<'a> {
    values: HashMap<String, Value>,
    parent: Option<&'a Environment<'a>>,
}

impl<'a> Environment<'a> {
    pub fn new(parent: Option<&'a Environment<'a>>) -> Environment<'a> {
        Environment {
            values: HashMap::new(),
            parent,
        }
    }

    pub fn get(&self, name: &str) -> Result<&Value, RuntimeError> {
        self.values.get(name).map_or_else(
            || {
                self.parent
                    .map(|parent| parent.get(name))
                    .unwrap_or(Err(RuntimeError::UndefinedVariable(name.to_string())))
            },
            Ok,
        )
    }

    pub fn define(&mut self, name: String, value: Value) {
        self.values.insert(name, value);
    }

    pub fn update(&mut self, name: &str, value: Value) -> Result<(), RuntimeError> {
        match self.values.get_mut(name) {
            None => return Err(RuntimeError::UndefinedVariable(name.to_string())),
            Some(existed) => {
                *existed = value;
                Ok(())
            }
        }
    }
}

#[cfg(test)]
mod test {
    use crate::environment::Environment;
    use crate::value::Value;

    #[test]
    fn nested() {
        let mut global = Environment::new(None);
        global.define("a".to_string(), Value::Number(1.0));
        let mut outer = Environment::new(Some(&global));
        outer.define("b".to_string(), Value::Number(2.0));
        outer.define("c".to_string(), Value::Number(0.0));
        let mut inner = Environment::new(Some(&outer));
        inner.define("c".to_string(), Value::Number(3.0)); // shadow c

        assert_eq!(inner.get("a").unwrap().as_number().unwrap(), 1.0);
        assert_eq!(inner.get("b").unwrap().as_number().unwrap(), 2.0);
        assert_eq!(inner.get("c").unwrap().as_number().unwrap(), 3.0);
        assert_eq!(outer.get("c").unwrap().as_number().unwrap(), 0.0);
    }
}
