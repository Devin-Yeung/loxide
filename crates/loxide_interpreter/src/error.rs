use crate::value::{Value, ValueKind};
use loxide_parser::token::Span;
use miette::Diagnostic;
use thiserror::Error;

#[derive(Debug, Error, Diagnostic)]
#[non_exhaustive]
pub enum RuntimeError {
    #[error("Invalid unary operand, expected: {1}")]
    InvalidUnaryOperand(#[label("expect a {1} type here")] Span, &'static str),
    #[error("Invalid binary operand, expected: {1}")]
    InvalidBinaryOperand(#[label("expect a {1} type here")] Span, &'static str),
    #[error("Incompatible operand, {lhs} is incompatible with {rhs}")]
    IncompatibleBinaryOperand {
        #[label("incompatible operand")]
        span: Span,
        lhs: ValueKind,
        rhs: ValueKind,
    },
    #[error("Undefined Variable: {0}")]
    UndefinedVariable(String),
    #[error("Expected boolean in condition expression")]
    ExpectedBoolean,
    #[error("Return value")]
    ReturnValue(Value),
    #[error("Bad arity, expected: {expected}, found: {found}")]
    BadArity { expected: usize, found: usize },
    #[error("Function call on a non-callable value")]
    CallOnNonCallable,
}
