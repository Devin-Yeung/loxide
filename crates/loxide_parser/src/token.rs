use crate::ast;
use crate::ast::{BinaryOperator, UnaryOperator};
use crate::error::SyntaxError;
use miette::SourceSpan;
use std::borrow::Cow;
use std::str::FromStr;

#[derive(Debug, PartialEq, Clone)]
pub enum Literal<'src> {
    String(Cow<'src, str>),
    Number(f64),
}

#[derive(Debug, PartialEq, Copy, Clone)]
pub enum Keyword {
    And,
    Class,
    Else,
    False,
    Fun,
    For,
    If,
    Nil,
    Or,
    Print,
    Return,
    Super,
    This,
    True,
    Var,
    While,
}

impl FromStr for Keyword {
    type Err = crate::error::SyntaxError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "and" => Ok(Keyword::And),
            "class" => Ok(Keyword::Class),
            "else" => Ok(Keyword::Else),
            "false" => Ok(Keyword::False),
            "fun" => Ok(Keyword::Fun),
            "for" => Ok(Keyword::For),
            "if" => Ok(Keyword::If),
            "nil" => Ok(Keyword::Nil),
            "or" => Ok(Keyword::Or),
            "print" => Ok(Keyword::Print),
            "return" => Ok(Keyword::Return),
            "super" => Ok(Keyword::Super),
            "this" => Ok(Keyword::This),
            "true" => Ok(Keyword::True),
            "var" => Ok(Keyword::Var),
            "while" => Ok(Keyword::While),
            _ => Err(crate::error::SyntaxError::InvalidKeywords),
        }
    }
}

#[derive(Debug, PartialEq, Clone)]
pub enum TokenType<'src> {
    // Single-character tokens.
    /// Opening parentheses `(`
    LeftParen,
    /// Closing parentheses `)`
    RightParen,
    /// Opening curly brace `{`
    LeftBrace,
    /// Closing curly brace `}`
    RightBrace,
    /// Comma `,`
    Comma,
    Dot,
    Minus,
    Plus,
    Semicolon,
    Slash,
    Star,
    // One or two character tokens.
    Bang,
    BangEqual,
    Equal,
    EqualEqual,
    Greater,
    GreaterEqual,
    Less,
    LessEqual,
    // Comment
    Comment,
    EOF,
    // Literals.
    Literal(Literal<'src>),
    // Keywords.
    Keyword(Keyword),
    Identifier,
}

impl<'src> TokenType<'src> {
    pub fn name(&self) -> &'static str {
        match self {
            TokenType::LeftParen => "(",
            TokenType::RightParen => ")",
            TokenType::LeftBrace => "{",
            TokenType::RightBrace => "}",
            TokenType::Comma => ",",
            TokenType::Dot => ".",
            TokenType::Minus => "-",
            TokenType::Plus => "+",
            TokenType::Semicolon => ";",
            TokenType::Slash => "/",
            TokenType::Star => "*",
            TokenType::Bang => "!",
            TokenType::BangEqual => "!=",
            TokenType::Equal => "=",
            TokenType::EqualEqual => "==",
            TokenType::Greater => ">",
            TokenType::GreaterEqual => ">=",
            TokenType::Less => "<",
            TokenType::LessEqual => "<=",
            TokenType::Comment => "comment",
            TokenType::EOF => "eof",
            TokenType::Identifier => "identifier",
            TokenType::Literal(_) => "literal",
            TokenType::Keyword(_) => "keyword",
        }
    }
}

#[derive(Debug)]
pub struct Token<'src> {
    pub(crate) ty: TokenType<'src>,
    pub(crate) lexeme: &'src str,
    pub(crate) span: Span,
}

/// The span of a token, indexed from 0, left inclusive, right exclusive.
#[derive(Debug, Copy, Clone, Eq, PartialEq)]
pub struct Span {
    pub start: usize,
    pub end: usize,
}

impl From<Span> for (usize, usize) {
    fn from(val: Span) -> Self {
        (val.start, val.end)
    }
}

impl Span {
    pub fn new(start: usize, end: usize) -> Self {
        Span { start, end }
    }
}

impl From<Span> for SourceSpan {
    fn from(val: Span) -> Self {
        SourceSpan::from(val.start..val.end)
    }
}

impl<'src> Token<'src> {
    pub fn new(ty: TokenType<'src>, lexeme: &'src str, span: Span) -> Token<'src> {
        Token { ty, lexeme, span }
    }
}

impl<'src> TryInto<BinaryOperator> for Token<'src> {
    type Error = SyntaxError;

    fn try_into(self) -> Result<BinaryOperator, Self::Error> {
        let op = match self.ty {
            TokenType::EqualEqual => BinaryOperator::EqualEqual,
            TokenType::BangEqual => BinaryOperator::BangEq,
            TokenType::GreaterEqual => BinaryOperator::GreaterEqual,
            TokenType::Greater => BinaryOperator::Greater,
            TokenType::Less => BinaryOperator::Less,
            TokenType::LessEqual => BinaryOperator::LessEqual,
            TokenType::Minus => BinaryOperator::Minus,
            TokenType::Plus => BinaryOperator::Plus,
            TokenType::Slash => BinaryOperator::Slash,
            TokenType::Star => BinaryOperator::Star,
            _ => return Err(SyntaxError::InvalidConversion),
        };
        Ok(op)
    }
}

impl<'src> TryInto<UnaryOperator> for Token<'src> {
    type Error = SyntaxError;

    fn try_into(self) -> Result<UnaryOperator, Self::Error> {
        let op = match self.ty {
            TokenType::Minus => UnaryOperator::Minus,
            TokenType::Bang => UnaryOperator::Bang,
            _ => return Err(SyntaxError::InvalidConversion),
        };
        Ok(op)
    }
}

impl<'src> TryInto<ast::Literal> for Token<'src> {
    type Error = SyntaxError;

    fn try_into(self) -> Result<ast::Literal, Self::Error> {
        let literal = match self.ty {
            TokenType::Literal(Literal::String(raw)) => ast::Literal::String(raw.to_string()),
            TokenType::Literal(Literal::Number(raw)) => ast::Literal::Number(raw),
            TokenType::Keyword(Keyword::Nil) => ast::Literal::Nil,
            TokenType::Keyword(Keyword::True) => ast::Literal::Boolean(true),
            TokenType::Keyword(Keyword::False) => ast::Literal::Boolean(false),
            _ => return Err(SyntaxError::InvalidConversion),
        };
        Ok(literal)
    }
}
