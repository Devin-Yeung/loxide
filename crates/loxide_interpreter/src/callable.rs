use crate::environment::Environment;
use crate::error::PrivateRuntimeError::Transparent;
use crate::error::{PrivateRuntimeError, RuntimeError};
use crate::eval::Evaluable;
use crate::value::Value;
use loxide_parser::ast::FunDeclaration;
use loxide_parser::token::Span;
use std::io::Write;
use std::sync::Arc;

#[derive(Debug, Clone)]
#[non_exhaustive]
pub enum Callable {
    Function(LoxFunction),
}

impl PartialEq for Callable {
    fn eq(&self, _other: &Self) -> bool {
        todo!()
    }
}

impl Callable {
    pub(crate) fn call(
        &self,
        arguments: Vec<Value>,
        env: &mut Environment,
        stdout: &mut impl Write,
    ) -> Result<Value, PrivateRuntimeError> {
        match self {
            Callable::Function(func) => func.call(arguments, env, stdout),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct LoxFunction {
    // we don't want to introduce any lifetime here,
    // since it will propagate to `Value`
    declaration: Arc<FunDeclaration>,
}

impl LoxFunction {
    pub fn new(declaration: Arc<FunDeclaration>) -> LoxFunction {
        LoxFunction { declaration }
    }

    pub(crate) fn call(
        &self,
        arguments: Vec<Value>,
        env: &mut Environment,
        stdout: &mut impl Write,
    ) -> Result<Value, PrivateRuntimeError> {
        let mut env = env.extend();

        if self.declaration.params.len() != arguments.len() {
            return Err(PrivateRuntimeError::BadArity {
                expected: self.declaration.params.len(),
                found: arguments.len(),
            });
        }

        // binding args to the environment
        for (value, arg) in arguments.into_iter().zip(&self.declaration.params) {
            env.define(arg, value);
        }

        for stmt in &self.declaration.body {
            match stmt.eval(&mut env, stdout) {
                Ok(_) => continue,
                Err(RuntimeError::ReturnValue(ret)) => return Ok(ret),
                Err(e) => return Err(Transparent(e)),
            }
        }

        Ok(Value::Void)
    }
}

impl Callable {
    pub fn function(declaration: Arc<FunDeclaration>) -> Callable {
        Callable::Function(LoxFunction::new(declaration))
    }

    pub fn params_span(&self) -> Span {
        match self {
            Callable::Function(func) => func.declaration.paren_token,
        }
    }
}
