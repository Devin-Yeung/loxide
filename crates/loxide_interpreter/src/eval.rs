use crate::callable::Callable;
use crate::environment::Environment;
use crate::error::RuntimeError;
use crate::value::{Value, ValueKind};
use loxide_parser::ast::{
    AssignExpr, BinaryExpr, BinaryOperator, CallExpr, ConditionStmt, Expr, ExprKind, ForStmt,
    GroupedExpr, Identifier, Literal, ReturnStmt, Stmt, UnaryExpr, UnaryOperator, WhileStmt,
};
use loxide_parser::token::Span;
use loxide_testsuite::probe;

pub trait Evaluable {
    fn eval(&self, env: &mut Environment) -> Result<Value, RuntimeError>;
}

macro_rules! inner_or {
    ($val:expr, $ty:tt, $err:expr) => {{
        ::paste::paste! {
            if $val.[<is_$ty>]() {
                Ok($val.[<as_$ty>]().unwrap())
            } else {
                Err($err)
            }
        }
    }};
}

impl Evaluable for Stmt {
    fn eval(&self, env: &mut Environment) -> Result<Value, RuntimeError> {
        match self {
            Stmt::Expression(e) => e.eval(env),
            Stmt::PrintStmt(e) => {
                probe!(e.eval(env))?;
                Ok(Value::Void)
            }
            Stmt::VarDeclaration(var, expr) => {
                let val = expr.eval(env)?;
                env.define(var.name.to_string(), val);
                Ok(Value::Void)
            }
            Stmt::FunDeclaration(decl) => {
                let callable = Callable::function(decl.clone());
                env.define(decl.name.name.to_string(), Value::Callable(callable));
                Ok(Value::Void)
            }
            Stmt::Block(stmts) => {
                // TODO: find a way to test the correctness of this evaluation
                let mut scope = env.extend();
                for stmt in stmts {
                    stmt.eval(&mut scope)?;
                }
                Ok(Value::Void)
            }
            Stmt::Condition(cond) => cond.eval(env),
            Stmt::While(cond) => cond.eval(env),
            Stmt::For(stmt) => stmt.eval(env),
            Stmt::ReturnStmt(stmt) => stmt.eval(env),
            _ => unreachable!(),
        }
    }
}

impl Evaluable for ConditionStmt {
    fn eval(&self, env: &mut Environment) -> Result<Value, RuntimeError> {
        let mut env = env.extend();
        match self.condition.eval(&mut env)? {
            Value::Boolean(b) => {
                if b {
                    self.then_branch.eval(&mut env)
                } else {
                    match &self.else_branch {
                        None => Ok(Value::Void),
                        Some(stmt) => stmt.eval(&mut env),
                    }
                }
            }
            _ => Err(RuntimeError::ExpectedBoolean),
        }
    }
}

impl Evaluable for WhileStmt {
    fn eval(&self, env: &mut Environment) -> Result<Value, RuntimeError> {
        while let Value::Boolean(b) = self.condition.eval(env)? {
            if b {
                self.body.eval(env)?;
            } else {
                return Ok(Value::Void);
            }
        }
        Err(RuntimeError::ExpectedBoolean)
    }
}

impl Evaluable for ForStmt {
    fn eval(&self, env: &mut Environment) -> Result<Value, RuntimeError> {
        let mut env = env.extend();
        let _ = &self.initializer.eval(&mut env)?;
        match &self.condition {
            None => {
                // infinite loop
                loop {
                    self.body.eval(&mut env)?;
                    self.increment.eval(&mut env)?;
                }
            }
            Some(cond) => {
                while let Value::Boolean(b) = cond.eval(&mut env)? {
                    if b {
                        self.body.eval(&mut env)?;
                        self.increment.eval(&mut env)?;
                    } else {
                        return Ok(Value::Void);
                    }
                }
            }
        }
        Err(RuntimeError::ExpectedBoolean)
    }
}

impl Evaluable for ReturnStmt {
    fn eval(&self, env: &mut Environment) -> Result<Value, RuntimeError> {
        self.value
            .as_ref()
            .map_or(Err(RuntimeError::ReturnValue(Value::Void)), |expr| {
                let value = expr.eval(env)?;
                Err(RuntimeError::ReturnValue(value))
            })
    }
}

impl Evaluable for Option<Box<Stmt>> {
    fn eval(&self, env: &mut Environment) -> Result<Value, RuntimeError> {
        if let Some(stmt) = self {
            stmt.eval(env)
        } else {
            Ok(Value::Void)
        }
    }
}

impl Evaluable for Option<Box<Expr>> {
    fn eval(&self, env: &mut Environment) -> Result<Value, RuntimeError> {
        let val = match self {
            None => Value::Nil,
            Some(e) => e.eval(env)?,
        };
        Ok(val)
    }
}

impl Evaluable for Option<Expr> {
    fn eval(&self, env: &mut Environment) -> Result<Value, RuntimeError> {
        let val = match self {
            None => Value::Nil,
            Some(e) => e.eval(env)?,
        };
        Ok(val)
    }
}

impl Evaluable for Expr {
    fn eval(&self, env: &mut Environment) -> Result<Value, RuntimeError> {
        match &self.kind {
            ExprKind::Literal(l) => l.eval(env),
            ExprKind::Unary(u) => u.eval(env),
            ExprKind::Binary(b) => b.eval(env),
            ExprKind::Grouped(g) => g.eval(env),
            ExprKind::Variable(v) => v.eval(env),
            ExprKind::Assign(a) => a.eval(env),
            ExprKind::Call(c) => c.eval(env),
            _ => unreachable!(),
        }
    }
}

impl Evaluable for CallExpr {
    fn eval(&self, env: &mut Environment) -> Result<Value, RuntimeError> {
        let callee = self.callee.eval(env)?;
        let args = self
            .args
            .iter()
            .map(|expr| expr.eval(env))
            .collect::<Result<Vec<Value>, RuntimeError>>()?;

        if let Value::Callable(callable) = callee {
            return callable.call(args, env);
        }

        Err(RuntimeError::CallOnNonCallable)
    }
}

impl Evaluable for AssignExpr {
    fn eval(&self, env: &mut Environment) -> Result<Value, RuntimeError> {
        let val = self.value.eval(env)?;
        env.mutate(&self.name.name, val.clone())?;
        Ok(val)
    }
}

impl Evaluable for BinaryExpr {
    fn eval(&self, env: &mut Environment) -> Result<Value, RuntimeError> {
        let lhs = self.lhs.eval(env)?;
        let rhs = self.rhs.eval(env)?;
        let val = match &self.operator {
            op @ (BinaryOperator::EqualEqual | BinaryOperator::BangEq) => {
                return if lhs.kind() == rhs.kind()
                    && (lhs.kind() == ValueKind::Number
                        || lhs.kind() == ValueKind::Boolean
                        || lhs.kind() == ValueKind::String)
                {
                    match op {
                        BinaryOperator::EqualEqual => Ok(Value::Boolean(lhs == rhs)),
                        BinaryOperator::BangEq => Ok(Value::Boolean(lhs != rhs)),
                        _ => unreachable!(),
                    }
                } else {
                    Err(RuntimeError::IncompatibleBinaryOperand {
                        span: Span::new(self.lhs.span().start, self.rhs.span().end),
                        lhs: lhs.kind(),
                        rhs: rhs.kind(),
                    })
                }
            }
            BinaryOperator::GreaterEqual => {
                let lhs = inner_or!(
                    lhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.lhs.span(), "number")
                )?;
                let rhs = inner_or!(
                    rhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.rhs.span(), "number")
                )?;
                Value::Boolean(lhs >= rhs)
            }
            BinaryOperator::Greater => {
                let lhs = inner_or!(
                    lhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.lhs.span(), "number")
                )?;
                let rhs = inner_or!(
                    rhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.rhs.span(), "number")
                )?;
                Value::Boolean(lhs > rhs)
            }
            BinaryOperator::Less => {
                let lhs = inner_or!(
                    lhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.lhs.span(), "number")
                )?;
                let rhs = inner_or!(
                    rhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.rhs.span(), "number")
                )?;
                Value::Boolean(lhs < rhs)
            }
            BinaryOperator::LessEqual => {
                let lhs = inner_or!(
                    lhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.lhs.span(), "number")
                )?;
                let rhs = inner_or!(
                    rhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.rhs.span(), "number")
                )?;
                Value::Boolean(lhs <= rhs)
            }
            BinaryOperator::Minus => {
                let lhs = inner_or!(
                    lhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.lhs.span(), "number")
                )?;
                let rhs = inner_or!(
                    rhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.rhs.span(), "number")
                )?;
                Value::Number(lhs - rhs)
            }
            BinaryOperator::Plus => {
                let lhs = inner_or!(
                    lhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.lhs.span(), "number")
                )?;
                let rhs = inner_or!(
                    rhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.rhs.span(), "number")
                )?;
                Value::Number(lhs + rhs)
            }
            BinaryOperator::Slash => {
                let lhs = inner_or!(
                    lhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.lhs.span(), "number")
                )?;
                let rhs = inner_or!(
                    rhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.rhs.span(), "number")
                )?;
                Value::Number(lhs / rhs)
            }
            BinaryOperator::Star => {
                let lhs = inner_or!(
                    lhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.lhs.span(), "number")
                )?;
                let rhs = inner_or!(
                    rhs,
                    number,
                    RuntimeError::InvalidBinaryOperand(self.rhs.span(), "number")
                )?;
                Value::Number(lhs * rhs)
            }
        };
        Ok(val)
    }
}

impl Evaluable for UnaryExpr {
    fn eval(&self, env: &mut Environment) -> Result<Value, RuntimeError> {
        let value = self.expr.eval(env)?;
        match self.operator {
            UnaryOperator::Minus => {
                let num = inner_or!(
                    value,
                    number,
                    RuntimeError::InvalidUnaryOperand(self.expr.span(), "number")
                )?;
                Ok(Value::Number(-num))
            }
            UnaryOperator::Bang => {
                let bool = inner_or!(
                    value,
                    boolean,
                    RuntimeError::InvalidUnaryOperand(self.expr.span(), "boolean")
                )?;
                Ok(Value::Boolean(!bool))
            }
        }
    }
}

impl Evaluable for GroupedExpr {
    fn eval(&self, env: &mut Environment) -> Result<Value, RuntimeError> {
        self.expr.eval(env)
    }
}

impl Evaluable for Identifier {
    fn eval(&self, env: &mut Environment) -> Result<Value, RuntimeError> {
        env.get(&self.name)
    }
}

impl Evaluable for Literal {
    fn eval(&self, _env: &mut Environment) -> Result<Value, RuntimeError> {
        let value = match *self {
            Literal::String(ref v) => Value::String(v.to_string()),
            Literal::Number(v) => Value::Number(v),
            Literal::Boolean(v) => Value::Boolean(v),
            Literal::Nil => Value::Nil,
        };
        Ok(value)
    }
}

#[cfg(test)]
mod tests {
    use crate::environment::Environment;
    use crate::error::RuntimeError;
    use crate::eval::Evaluable;
    use crate::value::Value;
    use loxide_diagnostic::reporter::{Reporter, Style};
    use loxide_testsuite::{footprints, register, unittest};
    use miette::Report;
    use std::sync::Arc;

    fn eval(src: &str) -> Vec<Result<Value, RuntimeError>> {
        let mut parser = loxide_parser::parser::Parser::new(src);
        let mut env = Environment::global();
        let (stmts, _) = parser.parse();
        stmts.into_iter().map(|stmt| stmt.eval(&mut env)).collect()
    }

    fn display_eval_error(src: &str) -> String {
        let mut parser = loxide_parser::parser::Parser::new(src);
        let mut env = Environment::global();
        let source = Arc::new(src.to_string());

        let (stmts, _) = parser.parse();
        stmts
            .into_iter()
            .map(|stmt| stmt.eval(&mut env))
            .filter(Result::is_err)
            .map(Result::unwrap_err)
            .map(|err| {
                let mut reporter = Reporter::new(Style::NoColor);
                let report: Report = err.into();
                reporter.push(report.with_source_code(source.clone()));
                reporter.report_to_string()
            })
            .collect::<Vec<_>>()
            .join("\n")
    }

    unittest!(literal, |src| {
        let results = eval(src);
        insta::assert_debug_snapshot!(results);
    });

    unittest!(valid_unary, |src| {
        let results = eval(src);
        insta::assert_debug_snapshot!(results);
    });

    unittest!(invalid_unary, |src| {
        let results = src
            .split('\n')
            .map(display_eval_error)
            .collect::<Vec<_>>()
            .join("\n");
        insta::assert_snapshot!(results);
    });

    unittest!(invalid_binary, |src| {
        let results = src
            .split('\n')
            .map(display_eval_error)
            .collect::<Vec<_>>()
            .join("\n");
        insta::assert_snapshot!(results);
    });

    unittest!(valid_binary, |src| {
        let results = eval(src);
        insta::assert_debug_snapshot!(results);
    });

    unittest!(variable, |src| {
        let results: Vec<_> = eval(src);
        insta::assert_debug_snapshot!(results);
    });

    unittest!(block_stmt, |src| {
        register!();
        eval(src);
        insta::assert_debug_snapshot!(footprints!());
    });

    unittest!(if_stmt, |src| {
        register!();
        eval(src);
        insta::assert_debug_snapshot!(footprints!());
    });

    unittest!(while_stmt, |src| {
        register!();
        eval(src);
        insta::assert_debug_snapshot!(footprints!());
    });

    unittest!(for_stmt, |src| {
        register!();
        eval(src);
        insta::assert_debug_snapshot!(footprints!());
    });

    unittest!(for_stmt_expect_bool, |src| {
        let results: Vec<_> = eval(src);
        insta::assert_debug_snapshot!(results);
    });

    unittest!(fn_call, |src| {
        register!();
        eval(src);
        insta::assert_debug_snapshot!(footprints!());
    });

    unittest!(bad_fn_call, |src| {
        let results: Vec<_> = eval(src);
        insta::assert_debug_snapshot!(results);
    });
}
