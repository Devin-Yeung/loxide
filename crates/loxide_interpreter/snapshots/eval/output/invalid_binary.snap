---
source: crates/loxide_interpreter/src/eval.rs
expression: results
---
  × Invalid binary operand, expected: number
   ╭────
 1 │ true + 1;
   · ──┬─
   ·   ╰── expect a number type here
   ╰────

  × Incompatible operand, boolean is incompatible with number
   ╭────
 1 │ true != 1;
   · ────┬────
   ·     ╰── incompatible operand
   ╰────
